plugins/outputs/sql/sql.go

```diff
@@ -201,28 +211,95 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
 		strings.Join(placeholders, ","))
 }
 
+func (p *SQL) createTable(metric telegraf.Metric) error {
+	tablename := metric.Name()
+	if _, found := p.tables[tablename]; !found && !p.tableExists(tablename) {
+		stmt := p.generateCreateTable(metric)
+		if _, err := p.db.Exec(stmt); err != nil {
+			return fmt.Errorf("creating table failed: %w", err)
+		}
+		p.tables[tablename] = make(map[string]bool)
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	return nil
+}
+
+func (p *SQL) createColumn(tablename, column, columnType string) error {
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		return nil
+	}
+
+	// update table cache with the latest table schema
+	// some servers may lack support for "IF NOT EXIST"
+	// clause in which case, we don't want to burden users
+	// with writing complex "IF NOT EXIST" workarounds
+	if !p.tables[tablename][column] {
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+
+	if !p.tables[tablename][column] {
+		createColumn := p.generateAddColumn(tablename, column, columnType)
+		if _, err := p.db.Exec(createColumn); err != nil {
+			return fmt.Errorf("creating column failed: %w", err)
+		}
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+
+	return nil
+}
+
 func (p *SQL) tableExists(tableName string) bool {
 	stmt := strings.ReplaceAll(p.TableExistsTemplate, "{TABLE}", quoteIdent(tableName))
 
 	_, err := p.db.Exec(stmt)
 	return err == nil
 }
 
+func (p *SQL) updateTableCache(tablename string) error {
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		return nil
+	}
+
+	stmt := strings.ReplaceAll(p.TableListColumnsTemplate, "{TABLE}", quoteStr(tablename))
+
+	columns, err := p.db.Query(stmt)
+	if err != nil {
+		return fmt.Errorf("fetching columns for table(%s) failed: %w", tablename, err)
+	}
+	defer columns.Close()
+
+	for columns.Next() {
+		var columnName string
+		if err := columns.Scan(&columnName); err != nil {
+			return err
+		}
+
+		if !p.tables[tablename][columnName] {
+			p.tables[tablename][columnName] = true
+		}
+	}
+
+	return nil
+}
+
 func (p *SQL) Write(metrics []telegraf.Metric) error {
 	var err error
 
 	for _, metric := range metrics {
 		tablename := metric.Name()
 
 		// create table if needed
-		if !p.tables[tablename] && !p.tableExists(tablename) {
-			createStmt := p.generateCreateTable(metric)
-			_, err := p.db.Exec(createStmt)
-			if err != nil {
-				return err
-			}
+		if err := p.createTable(metric); err != nil {
+			return err
```

Comment:
To be honest, I would like to keep the _condition_ here as otherwise this reads like "always create the table"!

Reply:
I was trying to keep the Write loop easier to read and handle the minutia in the respective function.

Reply:
Yes, please revert back. My reasoning is that we can directly see in the write loop that tables are _only_ created if they do not exist. Jumping to the `createTable` function only to see it directly returns is disturbing when debugging code...